File: 1302. Deepest Leaves Sum/Solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        max_depth = -1
        res = 0

        def dfs(root: TreeNode, depth: int) -> None:
            nonlocal max_depth, res
            if not root:
                return
            if depth > max_depth:
                max_depth = depth
                res = root.val
            elif depth == max_depth:
                res += root.val
            dfs(root.left, depth + 1)
            dfs(root.right, depth + 1)

        dfs(root, 0)
        return res

File: 1302. Deepest Leaves Sum/Solution.c
struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

void dfs(struct TreeNode* root, int depth, int *max_depth, int *res)
{
    if (!root)
        return;
    if (depth > *max_depth) {
        *max_depth = depth;
        *res = root->val;
    } else if (depth == *max_depth)
        *res += root->val;
    dfs(root->left, depth + 1, max_depth, res);
    dfs(root->right, depth + 1, max_depth, res);
}

int deepestLeavesSum(struct TreeNode* root)
{
    int max_depth = -1, res = 0;
    dfs(root, 0, &max_depth, &res);
    return res;
}
